jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/IRabitTracker.java

```diff
@@ -0,0 +1,20 @@
+package ml.dmlc.xgboost4j.java;
+
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Interface for Rabit tracker implementations with three public methods:
+ *
+ *  - start(): Start the rabit tracker awaiting for worker connections.
+ *  - getWorkerEnvs(): Return the environment variables needed to initialize Rabit clients.
+ *  - waitFor(): Wait for the task execution by the worker nodes.
+ */
+public interface IRabitTracker {
+  Map<String, String> getWorkerEnvs();
+  boolean start();
+  // for Scala-implementation compatibility
+  boolean start(long timeout, TimeUnit unit);
+  int waitFor();
+  int waitFor(long timeout, TimeUnit unit);
```

Comment:
My suggestion is 

* move `val tracker = startTracker(nWorkers)` to `buildDistributedBoosters`
* check `trainingData`'s STORAGE_LEVEL, if it is not cached, cache `partitionedTrainingSet` in `buildDistributedBoosters`; if it is cached, leave partitionedTrainingSet to save memory
* enforce a Spark job over `partitionedTrainingSet` before moved `val tracker = startTracker(nWorkers)`
* unpersist `partitionedTrainingSet` after a successful training